app/src/main/java/io/apicurio/registry/storage/impl/jdbc/JdbcRegistryStorage.java

```diff
@@ -0,0 +1,646 @@
+                    "DELETE FROM Artifact a " +
+                            "WHERE a.artifactId = :artifact_id")
+                    .setParameter("artifact_id", artifactId)
+                    .executeUpdate();
+
+            if (res1.size() != affected)
+                throw new RegistryStorageException("Assertion failed when deleting artifact '" + artifactId + "'");
+
+            // delete meta data
+            entityManager.createQuery(
+                    "DELETE FROM MetaData m " +
+                            "WHERE m.artifactId = :artifact_id")
+                    .setParameter("artifact_id", artifactId)
+                    .executeUpdate();
+
+            // delete rules
+            entityManager.createQuery(
+                    "DELETE FROM Rule r " +
+                            "WHERE r.artifactId = :artifact_id")
+                    .setParameter("artifact_id", artifactId)
+                    .executeUpdate();
+
+            return new TreeSet<>(res1);
+
+        } catch (PersistenceException ex) {
+            throw new RegistryStorageException(ex);
+        }
+    }
+
+    @Override
+    @Transactional
+    public StoredArtifact getArtifact(String artifactId) throws ArtifactNotFoundException, RegistryStorageException {
+        try {
+            Artifact artifact = _getArtifact(artifactId);
+
+            return mapper.toStoredArtifact(artifact);
+
+        } catch (NoResultException ex) {
+            throw new ArtifactNotFoundException(artifactId, ex);
+        } catch (PersistenceException ex) {
+            throw new RegistryStorageException(ex);
+        }
+    }
+
+    @Override
+    @Transactional
+    public ArtifactMetaDataDto updateArtifact(String artifactId, ArtifactType artifactType, String content) throws ArtifactNotFoundException, RegistryStorageException {
+        try {
+            // TODO mostly duplicate - do we need this method?
+            long nextVersion = _getNextArtifactVersion(artifactId);
+
+            if (nextVersion == 1) {
+                throw new ArtifactNotFoundException(artifactId);
+            }
+
+            Artifact artifact = Artifact.builder()
+                    .artifactId(artifactId)
+                    .version(nextVersion)
+                    .content(content)
+                    .build();
+
+            entityManager.persist(artifact);
+
+            new MetaDataMapperUpdater()
+                    .update(MetaDataKeys.TYPE, artifactType.value())
+                    .persistUpdate(entityManager, artifactId, nextVersion);
+
+            return new MetaDataMapperUpdater()
+                    .update(MetaDataKeys.TYPE, artifactType.value())
+                    .persistUpdate(entityManager, artifactId, null)
+                    .update(artifact)
+                    .toArtifactMetaDataDto();
+
+        } catch (PersistenceException ex) {
+            throw new RegistryStorageException(ex);
+        }
+    }
+
+    @Override
+    @Transactional
```

Comment:
Mark tx read-only, if that's possible / an option.

Reply:
I didn't find any way to do that yet:(